### PR TITLE
feat: add daemon

### PR DESCRIPTION
**Pull request by Cifko**
Added daemon with only few endpoints.
I had to greatly reduce the endpoints compared to atoma-node.
I will add more, once we switch atoma-proxy to have multiple users.

### CODE REVIEW

File: atoma-daemon/src/daemon.rs [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+/// let daemon_state = DaemonState::new(/* ... */);
+/// let app = create_daemon_router(daemon_state);
+/// // Start the server with the configured router
+/// axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
+///     .serve(app.into_make_service())
+///     .await?;
+/// ```
+pub fn create_daemon_router(daemon_state: DaemonState) -> Router {
+    Router::new()
+        .route("/subscriptions/:id", get(get_node_subscriptions))
```
Comment (jorgeantonio21) [inline atoma-daemon/src/daemon.rs:157]: add a `subscriptions` endpoint too. Do we really want a subscription by id ?

File: atoma-daemon/src/daemon.rs [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+/// Retrieves all stacks that are not settled.
+///
+/// # Arguments
+/// * `daemon_state` - The shared state containing the state manager
+///
+/// # Returns
+/// * `Result<Json<Vec<Stack>>>` - A JSON response containing a list of stacks
+///   - `Ok(Json<Vec<Stack>>)` - Successfully retrieved stacks
+///   - `Err(StatusCode::INTERNAL_SERVER_ERROR)` - Failed to retrieve stacks from state manager
+async fn get_current_stacks(State(daemon_state): State<DaemonState>) -> Result<Json<Vec<Stack>>> {
```
Comment (jorgeantonio21) [inline atoma-daemon/src/daemon.rs:175]: Add utoipa support pls

File: atoma-daemon/src/daemon.rs [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+///     {
+///         "node_small_id": 123,
+///         "model_name": "example_model",
+///         "echelon_id": 1,
+///         "subscription_time": "2024-03-21T12:00:00Z"
+///     }
+/// ]
+/// ```
+#[instrument(level = "trace", skip_all)]
+async fn get_node_subscriptions(
```
Comment (jorgeantonio21) [inline atoma-daemon/src/daemon.rs:212]: Add utoipa support

File: atoma-daemon/src/daemon.rs [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+/// [
+///    {
+///       "node_small_id": 123,
+///       "model_name": "example_model",
+///       "echelon_id": 1,
+///       "subscription_time": "2024-03-21T12:00:00Z"
+/// }
+/// ]
+/// ```
+async fn get_nodes_for_tasks(
```
Comment (jorgeantonio21) [inline atoma-daemon/src/daemon.rs:251]: Add instrumentation and utoipa

File: atoma-daemon/src/daemon.rs (hunk 2 above)
Comment (jorgeantonio21) [reply atoma-daemon/src/daemon.rs:175]: and instrumentation

File: atoma-daemon/src/daemon.rs [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+///
+/// # Returns
+/// * `Result<Json<Vec<Task>>>` - A JSON response containing a list of tasks
+///   - `Ok(Json<Vec<Task>>)` - Successfully retrieved tasks
+///   - `Err(StatusCode::INTERNAL_SERVER_ERROR)` - Failed to retrieve tasks from state manager
+///
+/// # Example Response
+/// Returns a JSON array of Task objects representing all tasks in the system
+#[instrument(level = "trace", skip_all)]
+async fn get_all_tasks(State(daemon_state): State<DaemonState>) -> Result<Json<Vec<Task>>> {
```
Comment (jorgeantonio21) [inline atoma-daemon/src/daemon.rs:279]: add utoipa

File: atoma-daemon/src/daemon.rs [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+        })?;
+    Ok(Json(all_tasks))
+}
+
+/// Health check endpoint for the daemon.
+///
+/// # Returns
+/// * `StatusCode::OK` - Always returns OK
+#[instrument(level = "trace", skip_all)]
+async fn health() -> StatusCode {
```
Comment (jorgeantonio21) [inline atoma-daemon/src/daemon.rs:296]: add utoipa

File: atoma-daemon/src/daemon.rs [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+///
+/// # Returns
+/// * `Result<Json<Vec<Stack>>>` - A JSON response containing a list of stacks
+///   - `Ok(Json<Vec<Stack>>)` - Successfully retrieved stacks
+///   - `Err(StatusCode::INTERNAL_SERVER_ERROR)` - Failed to retrieve stacks from state manager
+///
+/// # Example Response
+/// Returns a JSON array of Stack objects for the specified node
+#[instrument(level = "trace", skip_all)]
+async fn get_node_stacks(
```
Comment (jorgeantonio21) [inline atoma-daemon/src/daemon.rs:314]: add utoipa

File: atoma-proxy/src/main.rs [hunk 8]
```diff
@@ -161,6 +166,23 @@ async fn main() -> Result<()> {
...
 
+    let daemon_tcp_listener = TcpListener::bind(&config.daemon.service_bind_address)
+        .await
+        .context("Failed to bind daemon TCP listener")?;
+
+    let daemon_app_state = DaemonState {
+        atoma_state: AtomaState::new_from_url(&config.state.database_url).await?,
+    };
+
+    let daemon_handle = spawn_with_shutdown(
```
Comment (jorgeantonio21) [inline atoma-proxy/src/main.rs:177]: I would rename the service to something else than `daemon`, maybe proxy service ?

File: atoma-state/src/state_manager.rs [hunk 9]
```diff
@@ -550,6 +550,84 @@ impl AtomaState {
...
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// use atoma_node::atoma_state::{AtomaStateManager, NodeSubscription};
+    ///
+    /// async fn get_all_node_subscriptions_for_task(state_manager: &AtomaStateManager, task_small_id: i64) -> Result<Vec<NodeSubscription>, AtomaStateManagerError> {
+    ///    state_manager.get_all_node_subscriptions_for_task(task_small_id).await
+    /// }
+    /// ```
+    pub async fn get_all_node_subscriptions_for_task(
```
Comment (jorgeantonio21) [inline atoma-state/src/state_manager.rs:582]: add instrumentation

File: atoma-state/src/state_manager.rs [hunk 10]
```diff
@@ -550,6 +550,84 @@ impl AtomaState {
...
+    /// # Example
+    ///
+    /// ```rust,ignore
+    /// use atoma_node::atoma_state::AtomaStateManager;
+    ///
+    /// async fn get_current_stacks(state_manager: &AtomaStateManager) -> Result<Vec<Stack>, AtomaStateManagerError> {
+    ///    state_manager.get_current_stacks().await
+    /// }
+    /// ```
+    pub async fn get_current_stacks(&self) -> Result<Vec<Stack>> {
```
Comment (jorgeantonio21) [inline atoma-state/src/state_manager.rs:621]: add instrumentation

**Review by jorgeantonio21 — COMMENTED** [review]
Some changes are requested, these are outlined in the comments above.

File: atoma-daemon/src/daemon.rs (hunk 1 above)
Comment (Cifko) [reply atoma-daemon/src/daemon.rs:157]: changed to all subscriptions

File: atoma-proxy/src/main.rs (hunk 8 above)
Comment (Cifko) [reply atoma-proxy/src/main.rs:177]: done

**Review by jorgeantonio21 — APPROVED** [review]
LGTM

**Pull request merged**